### PR TITLE
Keycloak use HTTPS

### PR DESCRIPTION
**Pull request by blling**
Use HTTPS of keycloak.

- [x] Use `./script/start.sh` to start up all backgrand and test `ng client` buy something is ok
- [x] Use `docker-compose` to start up all backgrand and test `ng client` buy something is ok
- [x] Use `docker-compose` to start up all backgrand and use `java -jar` to start up gql-gateway and test `react client` without login is ok
- [x] Use `docker-compose` to start up all backgrand and use `java -jar` to start up gql-gateway and test `react client` in login status is NG, cause 500 error
Because of keycloak use has no email, add email to keycloak user this error is gone
- [x] Use `docker-compose` to start up all backgrand and use `java -jar` to start up gql-gateway and test `react client` buy something and the order is always `SUBMITTED`.
No notifacation, order status is always `SUBMITTED`, ***maybe the test data*** is incomplete, just make this test case pass. I use the following SQL to do data init:

```sql
insert into brand (id, name) values(gen_random_uuid(), 'brand 1');

insert into category (id, name) values(gen_random_uuid(), 'category 1');

insert into catalog_item (
id,
available_stock,
description,
name,
picture_file_name,
price,
brand_id,
category_id) values(
gen_random_uuid(),
1,
'describle 1',
'catalog_item 1',
'adidas-shoes-2.png',
43.5,
uuid('f2cbbe0e-37eb-4f50-a2a2-bd6167783cb3'),
uuid('5ea92515-cf6b-48a9-b3f5-d0aa0222df90')
);
```

> `brand_id` and `category_id` of catalog_item should be changed to the right one when `brand` and `catalog` inserted

There is a lot of test case i think, but i just do little ~~

### CODE REVIEW

**Comment by blling** [comment]
@hdimitrieski

**Comment by hdimitrieski** [comment]
@blling Thanks for the PR! :) I'll check it out by the end of the week.

File: docker/.env [hunk 1]
```diff
@@ -1,5 +1,5 @@
-CONFIG_SERVICE_USER=user
-CONFIG_SERVICE_PASSWORD=cfg-password
+CONFIG_SERVICE_USER=cfguser
```
Comment (blling) [inline docker/.env:1]: Use the same user and password with `dev` profile

File: README.md [hunk 2]
```diff
@@ -146,9 +165,17 @@ You can run all docker containers and Spring applications by running the `./scri
 won't start ELK and Zipkin. If you want to use ELK and Zipkin, you can run the script as:
 `./scripts/start.sh --elk --distributed-tracing`. To stop everything you can run `./scripts/stop.sh`.
 
+Do not forget to config the hosts file:
```
Comment (hdimitrieski) [inline README.md:168]: @blling Why do we have to configure the hosts file? Can we avoid this? If you don't have admin rights (like me :)) you can't actually change the hosts file.

**Comment by hdimitrieski** [comment]
@blling Can you please check the comment? :)

**Comment by blling** [comment]
> @blling Can you please check the comment? :)

I think I'm stuck by the hostname :-(

**Comment by hdimitrieski** [comment]
> 

@blling It's okay. I'll take a look at it. What exactly is the problem? Is setting **KC_HOSTNAME** to localhost, not allowed?

**Comment by blling** [comment]
> > 
> 
> @blling It's okay. I'll take a look at it. What exactly is the problem? Is setting **KC_HOSTNAME** to localhost, not allowed?

I think `KC_HOSTNAME` has nothing to do with `HTTPS`. Also **KC_HOSTNAME** should not be `localhost`, because in spring docker profile, `app.auth-server.host` is set to `authorization-service`. If **KC_HOSTNAME** is `localhost`, then the app run in docker container should not access keycloak on hostname `localhost`.  `localhost` is suitable for local development but not in docker container.

The problem for now is the `gateway` could not access keycloak server through `https://localhost:8443/auth/realms/e-shop/.well-known/openid-configuration`, the errorstack as follow:
```
Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2022-08-08 21:58:24.728 ERROR [api-gateway,,] 44810 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'analyticsApiServiceImpl' defined in URL [jar:file:/home/blling/Documents/Projects/e-shop/infrastructure/gateway/build/libs/gateway.jar!/BOOT-INF/classes!/com/eshop/gateway/infrastructure/AnalyticsApiServiceImpl.class]: Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'analyticsWebClient' defined in class path resource [com/eshop/gateway/config/AnalyticsWebClientConfig.class]: Unsatisfied dependency expressed through method 'analyticsWebClient' parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'clientRegistrationRepository' defined in class path resource [org/springframework/boot/autoconfigure/security/oauth2/client/reactive/ReactiveOAuth2ClientConfigurations$ReactiveClientRegistrationRepositoryConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.oauth2.client.registration.InMemoryReactiveClientRegistrationRepository]: Factory method 'clientRegistrationRepository' threw exception; nested exception is java.lang.IllegalArgumentException: Unable to resolve Configuration with the provided Issuer of "https://localhost:8443/auth/realms/e-shop"
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:800)
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:229)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1372)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1222)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:955)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:918)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583)
	at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:66)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:734)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:408)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:308)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1306)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1295)
	at com.eshop.gateway.GatewayApplication.main(GatewayApplication.java:13)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:108)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:65)
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'analyticsWebClient' defined in class path resource [com/eshop/gateway/config/AnalyticsWebClientConfig.class]: Unsatisfied dependency expressed through method 'analyticsWebClient' parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'clientRegistrationRepository' defined in class path resource [org/springframework/boot/autoconfigure/security/oauth2/client/reactive/ReactiveOAuth2ClientConfigurations$ReactiveClientRegistrationRepositoryConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.oauth2.client.registration.InMemoryReactiveClientRegistrationRepository]: Factory method 'clientRegistrationRepository' threw exception; nested exception is java.lang.IllegalArgumentException: Unable to resolve Configuration with the provided Issuer of "https://localhost:8443/auth/realms/e-shop"
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:800)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:541)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1391)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1311)
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:887)
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:791)
	... 27 common frames omitted
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'clientRegistrationRepository' defined in class path resource [org/springframework/boot/autoconfigure/security/oauth2/client/reactive/ReactiveOAuth2ClientConfigurations$ReactiveClientRegistrationRepositoryConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.oauth2.client.registration.InMemoryReactiveClientRegistrationRepository]: Factory method 'clientRegistrationRepository' threw exception; nested exception is java.lang.IllegalArgumentException: Unable to resolve Configuration with the provided Issuer of "https://localhost:8443/auth/realms/e-shop"
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:658)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:638)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1391)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1311)
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:887)
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:791)
	... 41 common frames omitted
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.security.oauth2.client.registration.InMemoryReactiveClientRegistrationRepository]: Factory method 'clientRegistrationRepository' threw exception; nested exception is java.lang.IllegalArgumentException: Unable to resolve Configuration with the provided Issuer of "https://localhost:8443/auth/realms/e-shop"
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653)
	... 55 common frames omitted
Caused by: java.lang.IllegalArgumentException: Unable to resolve Configuration with the provided Issuer of "https://localhost:8443/auth/realms/e-shop"
	at org.springframework.security.oauth2.client.registration.ClientRegistrations.getBuilder(ClientRegistrations.java:220)
	at org.springframework.security.oauth2.client.registration.ClientRegistrations.fromIssuerLocation(ClientRegistrations.java:144)
	at org.springframework.boot.autoconfigure.security.oauth2.client.OAuth2ClientPropertiesRegistrationAdapter.getBuilderFromIssuerIfPossible(OAuth2ClientPropertiesRegistrationAdapter.java:83)
	at org.springframework.boot.autoconfigure.security.oauth2.client.OAuth2ClientPropertiesRegistrationAdapter.getClientRegistration(OAuth2ClientPropertiesRegistrationAdapter.java:59)
	at org.springframework.boot.autoconfigure.security.oauth2.client.OAuth2ClientPropertiesRegistrationAdapter.lambda$getClientRegistrations$0(OAuth2ClientPropertiesRegistrationAdapter.java:53)
	at java.base/java.util.HashMap.forEach(HashMap.java:1421)
	at org.springframework.boot.autoconfigure.security.oauth2.client.OAuth2ClientPropertiesRegistrationAdapter.getClientRegistrations(OAuth2ClientPropertiesRegistrationAdapter.java:52)
	at org.springframework.boot.autoconfigure.security.oauth2.client.reactive.ReactiveOAuth2ClientConfigurations$ReactiveClientRegistrationRepositoryConfiguration.clientRegistrationRepository(ReactiveOAuth2ClientConfigurations.java:56)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
	... 56 common frames omitted
Caused by: org.springframework.web.client.ResourceAccessException: I/O error on GET request for "https://localhost:8443/auth/realms/e-shop/.well-known/openid-configuration": No name matching localhost found; nested exception is javax.net.ssl.SSLHandshakeException: No name matching localhost found
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:785)
	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:670)
	at org.springframework.security.oauth2.client.registration.ClientRegistrations.lambda$oidc$0(ClientRegistrations.java:155)
	at org.springframework.security.oauth2.client.registration.ClientRegistrations.getBuilder(ClientRegistrations.java:208)
	... 68 common frames omitted
Caused by: javax.net.ssl.SSLHandshakeException: No name matching localhost found
	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:131)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:371)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:314)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:309)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1357)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.onConsumeCertificate(CertificateMessage.java:1232)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.consume(CertificateMessage.java:1175)
	at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:396)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:480)
	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:458)
	at java.base/sun.security.ssl.TransportContext.dispatch(TransportContext.java:201)
	at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:172)
	at java.base/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1500)
	at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1415)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:450)
	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:421)
	at java.base/sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:580)
	at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:183)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:142)
	at org.springframework.http.client.SimpleBufferingClientHttpRequest.executeInternal(SimpleBufferingClientHttpRequest.java:76)
	at org.springframework.http.client.AbstractBufferingClientHttpRequest.executeInternal(AbstractBufferingClientHttpRequest.java:48)
	at org.springframework.http.client.AbstractClientHttpRequest.execute(AbstractClientHttpRequest.java:66)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:776)
	... 71 common frames omitted
Caused by: java.security.cert.CertificateException: No name matching localhost found
	at java.base/sun.security.util.HostnameChecker.matchDNS(HostnameChecker.java:234)
	at java.base/sun.security.util.HostnameChecker.match(HostnameChecker.java:103)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkIdentity(X509TrustManagerImpl.java:452)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkIdentity(X509TrustManagerImpl.java:426)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:238)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:132)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1341)
	... 89 common frames omitted
```

**Comment by hdimitrieski** [comment]
@blling We can use HTTPS only when we run the services with docker. The **docker** profile is set when running the services with docker. So we can configure the URL to the authorization service accordingly. Would that solve the problem?

**Comment by blling** [comment]
> @blling We can use HTTPS only when we run the services with docker. The **docker** profile is set when running the services with docker. So we can configure the URL to the authorization service accordingly. Would that solve the problem?

It works will when run all the services with docker for now.

**Comment by hdimitrieski** [comment]
> > @blling We can use HTTPS only when we run the services with docker. The **docker** profile is set when running the services with docker. So we can configure the URL to the authorization service accordingly. Would that solve the problem?
> 
> It works well when run all the services with docker for now.
> 
> Config the hostname in hosts file can solve all the problem…

@blling Yes, but if you don't have admin access, you can't configure the **hosts** file... Can we enable both HTTP and HTTPS, so that when running the services without docker we can use HTTP?

**Comment by blling** [comment]
> > > @blling We can use HTTPS only when we run the services with docker. The **docker** profile is set when running the services with docker. So we can configure the URL to the authorization service accordingly. Would that solve the problem?
> > 
> > 
> > It works well when run all the services with docker for now.
> > Config the hostname in hosts file can solve all the problem…
> 
> @blling Yes, but if you don't have admin access, you can't configure the **hosts** file... Can we enable both HTTP and HTTPS, so that when running the services without docker we can use HTTP?

I will give a try tomorrow:-)

**Comment by blling** [comment]
@hdimitrieski When enable the keycloak HTTP, now do not need to config hostname when all services start by dev profile.

**Comment by hdimitrieski** [comment]
> @hdimitrieski When enable the keycloak HTTP, now do not need to config hostname when all services start by dev profile.

That's great! Thanks for the PR! :)

**Pull request merged**